// raspi-i2c/lib/index.js
"use strict";

var _interopRequire = function (obj) { return obj && obj.__esModule ? obj["default"] : obj; };

var _defineProperty = function (obj, key, value) { return Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); };

var _createClass = (function () { function defineProperties(target, props) { for (var key in props) { var prop = props[key]; prop.configurable = true; if (prop.value) prop.writable = true; } Object.defineProperties(target, props); } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

/*
The MIT License (MIT)

Copyright (c) 2015 Bryan Hughes <bryan@theoreticalideations.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

var i2c = _interopRequire(require("i2c-bus"));

var execSync = require("child_process").execSync;

var Peripheral = require("raspi-peripheral").Peripheral;

var _raspiBoard = require("raspi-board");

var VERSION_1_MODEL_B_REV_1 = _raspiBoard.VERSION_1_MODEL_B_REV_1;
var getBoardRevision = _raspiBoard.getBoardRevision;

if (typeof execSync !== "function") {
  execSync = require("execSync").run;
}

function checkAddress(address) {
  if (typeof address !== "number" || address < 0 || address > 127) {
    throw new Error("Invalid I2C address " + address + ". Valid addresses are 0 through 0x7f.");
  }
}

function checkRegister(register) {
  if (register !== undefined && (typeof register !== "number" || register < 0 || register > 255)) {
    throw new Error("Invalid I2C register " + register + ". Valid registers are 0 through 0xff.");
  }
}

function checkLength(length) {
  if (typeof length !== "number" || length < 0 || length > 32) {
    throw new Error("Invalid I2C length " + length + ". Valid lengths are 0 through 32.");
  }
}

function checkCallback(cb) {
  if (typeof cb !== "function") {
    throw new Error("Invalid I2C callback " + cb);
  }
}

function checkBuffer(buffer) {
  if (!Buffer.isBuffer(buffer) || buffer.length <= 0 || buffer.length > 32) {
    throw new Error("Invalid I2C buffer " + buffer + ". Valid lengths are 0 through 32.");
  }
}

function checkByte(byte) {
  if (typeof byte !== "number" || byte < 0 || byte > 255) {
    throw new Error("Invalid I2C byte " + byte + ". Valid values are 0 through 0xff.");
  }
}

function checkWord(word) {
  if (typeof word !== "number" || word < 0 || word > 65535) {
    throw new Error("Invalid I2C word " + word + ". Valid values are 0 through 0xffff.");
  }
}

var devices = Symbol("devices");
var getDevice = Symbol("getDevice");

var I2C = exports.I2C = (function (_Peripheral) {
  function I2C(options) {
    _classCallCheck(this, I2C);

    var pins = options;
    if (!Array.isArray(pins)) {
      options = options || {};
      pins = options.pins || ["SDA0", "SCL0"];
    }
    _get(Object.getPrototypeOf(I2C.prototype), "constructor", this).call(this, pins);

    Object.defineProperties(this, _defineProperty({}, devices, {
      writable: true,
      value: []
    }));

    execSync("modprobe i2c-dev");
  }

  _inherits(I2C, _Peripheral);

  _createClass(I2C, (function () {
    var _createClass2 = {
      destroy: {
        value: function destroy() {
          this[devices].forEach(function (device) {
            device.closeSync();
          });

          this[devices] = [];

          _get(Object.getPrototypeOf(I2C.prototype), "destroy", this).call(this);
        }
      }
    };

    _defineProperty(_createClass2, getDevice, {
      value: function (address) {
        var device = this[devices][address];

        if (device === undefined) {
          device = i2c.openSync(getBoardRevision() === VERSION_1_MODEL_B_REV_1 ? 0 : 1);
          this[devices][address] = device;
        }

        return device;
      }
    });

    _defineProperty(_createClass2, "read", {
      value: function read(address, register, length, cb) {
        this.validateAlive();

        if (arguments.length === 3) {
          cb = length;
          length = register;
          register = undefined;
        }

        checkAddress(address);
        checkRegister(register);
        checkLength(length);
        checkCallback(cb);

        var buffer = new Buffer(length);
        function callback(err) {
          if (err) {
            return cb(err);
          }
          cb(null, buffer);
        }

        if (register === undefined) {
          this[getDevice](address).i2cRead(address, length, buffer, callback);
        } else {
          this[getDevice](address).readI2cBlock(address, register, length, buffer, callback);
        }
      }
    });

    _defineProperty(_createClass2, "readSync", {
      value: function readSync(address, register, length) {
        this.validateAlive();

        if (arguments.length === 2) {
          length = register;
          register = undefined;
        }

        checkAddress(address);
        checkRegister(register);
        checkLength(length);

        var buffer = new Buffer(length);

        if (register === undefined) {
          this[getDevice](address).i2cReadSync(address, length, buffer);
        } else {
          this[getDevice](address).readI2cBlockSync(address, register, length, buffer);
        }

        return buffer;
      }
    });

    _defineProperty(_createClass2, "readByte", {
      value: function readByte(address, register, cb) {
        var _this = this;

        this.validateAlive();

        if (arguments.length === 2) {
          cb = register;
          register = undefined;
        }

        checkAddress(address);
        checkRegister(register);
        checkCallback(cb);

        if (register === undefined) {
          (function () {
            var buffer = new Buffer(1);
            _this[getDevice](address).i2cRead(address, buffer.length, buffer, function (err) {
              if (err) {
                return cb(err);
              }
              cb(null, buffer[0]);
            });
          })();
        } else {
          this[getDevice](address).readByte(address, register, cb);
        }
      }
    });

    _defineProperty(_createClass2, "readByteSync", {
      value: function readByteSync(address, register) {
        this.validateAlive();

        checkAddress(address);
        checkRegister(register);

        var byte = undefined;
        if (register === undefined) {
          var buffer = new Buffer(1);
          this[getDevice](address).i2cReadSync(address, buffer.length, buffer);
          byte = buffer[0];
        } else {
          byte = this[getDevice](address).readByteSync(address, register);
        }
        return byte;
      }
    });

    _defineProperty(_createClass2, "readWord", {
      value: function readWord(address, register, cb) {
        var _this = this;

        this.validateAlive();

        if (arguments.length === 2) {
          cb = register;
          register = undefined;
        }

        checkAddress(address);
        checkRegister(register);
        checkCallback(cb);

        if (register === undefined) {
          (function () {
            var buffer = new Buffer(2);
            _this[getDevice](address).i2cRead(address, buffer.length, buffer, function (err) {
              if (err) {
                return cb(err);
              }
              cb(null, buffer.readUInt16LE(0));
            });
          })();
        } else {
          this[getDevice](address).readWord(address, register, cb);
        }
      }
    });

    _defineProperty(_createClass2, "readWordSync", {
      value: function readWordSync(address, register) {
        this.validateAlive();

        checkAddress(address);
        checkRegister(register);

        var byte = undefined;
        if (register === undefined) {
          var buffer = new Buffer(2);
          this[getDevice](address).i2cReadSync(address, buffer.length, buffer);
          byte = buffer.readUInt16LE(0);
        } else {
          byte = this[getDevice](address).readWordSync(address, register);
        }
        return byte;
      }
    });

    _defineProperty(_createClass2, "write", {
      value: function write(address, register, buffer, cb) {
        this.validateAlive();

        if (arguments.length === 3) {
          cb = buffer;
          buffer = register;
          register = undefined;
        }

        checkAddress(address);
        checkRegister(register);
        checkBuffer(buffer);
        checkCallback(cb);

        if (register === undefined) {
          this[getDevice](address).i2cWrite(address, buffer.length, buffer, cb);
        } else {
          this[getDevice](address).writeI2cBlock(address, register, buffer.length, buffer, cb);
        }
      }
    });

    _defineProperty(_createClass2, "writeSync", {
      value: function writeSync(address, register, buffer) {
        this.validateAlive();

        if (arguments.length === 2) {
          buffer = register;
          register = undefined;
        }

        checkAddress(address);
        checkRegister(register);
        checkBuffer(buffer);

        if (register === undefined) {
          this[getDevice](address).i2cWriteSync(address, buffer.length, buffer);
        } else {
          this[getDevice](address).writeI2cBlockSync(address, register, buffer.length, buffer);
        }
      }
    });

    _defineProperty(_createClass2, "writeByte", {
      value: function writeByte(address, register, byte, cb) {
        this.validateAlive();

        if (arguments.length === 3) {
          cb = byte;
          byte = register;
          register = undefined;
        }

        checkAddress(address);
        checkRegister(register);
        checkByte(byte);
        checkCallback(cb);

        if (register === undefined) {
          this[getDevice](address).i2cWrite(address, 1, new Buffer([byte]), cb);
        } else {
          this[getDevice](address).writeByte(address, register, byte, cb);
        }
      }
    });

    _defineProperty(_createClass2, "writeByteSync", {
      value: function writeByteSync(address, register, byte) {
        this.validateAlive();

        if (arguments.length === 2) {
          byte = register;
          register = undefined;
        }

        checkAddress(address);
        checkRegister(register);
        checkByte(byte);

        if (register === undefined) {
          this[getDevice](address).i2cWriteSync(address, 1, new Buffer([byte]));
        } else {
          this[getDevice](address).writeByteSync(address, register, byte);
        }
      }
    });

    _defineProperty(_createClass2, "writeWord", {
      value: function writeWord(address, register, word, cb) {
        this.validateAlive();

        if (arguments.length === 3) {
          cb = word;
          word = register;
          register = undefined;
        }

        checkAddress(address);
        checkRegister(register);
        checkWord(word);
        checkCallback(cb);

        if (register === undefined) {
          var buffer = new Buffer(2);
          buffer.writeUInt16LE(word, 0);
          this[getDevice](address).i2cWrite(address, buffer.length, buffer, cb);
        } else {
          this[getDevice](address).writeWord(address, register, word, cb);
        }
      }
    });

    _defineProperty(_createClass2, "writeWordSync", {
      value: function writeWordSync(address, register, word) {
        this.validateAlive();

        if (arguments.length === 2) {
          word = register;
          register = undefined;
        }

        checkAddress(address);
        checkRegister(register);
        checkWord(word);

        if (register === undefined) {
          var buffer = new Buffer(2);
          buffer.writeUInt16LE(word, 0);
          this[getDevice](address).i2cWriteSync(address, buffer.length, buffer);
        } else {
          this[getDevice](address).writeWordSync(address, register, word);
        }
      }
    });

    return _createClass2;
  })());

  return I2C;
})(Peripheral);

Object.defineProperty(exports, "__esModule", {
  value: true
});
