// raspi-gpio/lib/index.js
"use strict";

var _interopRequire = function (obj) { return obj && obj.__esModule ? obj["default"] : obj; };

var _prototypeProperties = function (child, staticProps, instanceProps) { if (staticProps) Object.defineProperties(child, staticProps); if (instanceProps) Object.defineProperties(child.prototype, instanceProps); };

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc && desc.writable) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

/*
The MIT License (MIT)

Copyright (c)2014 Bryan Hughes <bryan@theoreticalideations.com> (http://theoreticalideations.com)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

var Peripheral = require("raspi-peripheral").Peripheral;

var addon = _interopRequire(require("../build/Release/addon"));

var INPUT = 0;
var OUTPUT = 1;

var LOW = exports.LOW = 0;
var HIGH = exports.HIGH = 1;

var PULL_NONE = exports.PULL_NONE = 0;
var PULL_UP = exports.PULL_UP = 1;
var PULL_DOWN = exports.PULL_DOWN = 2;

function parseConfig(config) {
  var pin = undefined;
  var pullResistor = undefined;
  if (typeof config == "number" || typeof config == "string") {
    pin = config;
    pullResistor = PULL_NONE;
  } else if (typeof config == "object") {
    pin = config.pin;
    pullResistor = config.pullResistor || PULL_NONE;
    if ([PULL_NONE, PULL_DOWN, PULL_UP].indexOf(pullResistor) == -1) {
      throw new Error("Invalid pull resistor option " + pullResistor);
    }
  } else {
    throw new Error("Invalid pin or configuration");
  }
  return {
    pin: pin,
    pullResistor: pullResistor
  };
}

var DigitalOutput = exports.DigitalOutput = (function (Peripheral) {
  function DigitalOutput(config) {
    _classCallCheck(this, DigitalOutput);

    config = parseConfig(config);
    _get(Object.getPrototypeOf(DigitalOutput.prototype), "constructor", this).call(this, config.pin);
    addon.init(this.pins[0], config.pullResistor, OUTPUT);
  }

  _inherits(DigitalOutput, Peripheral);

  _prototypeProperties(DigitalOutput, null, {
    write: {
      value: function write(value) {
        if (!this.alive) {
          throw new Error("Attempted to write to a destroyed peripheral");
        }
        if ([LOW, HIGH].indexOf(value) == -1) {
          throw new Error("Invalid write value " + value);
        }
        addon.write(this.pins[0], value);
      },
      writable: true,
      configurable: true
    }
  });

  return DigitalOutput;
})(Peripheral);

var DigitalInput = exports.DigitalInput = (function (Peripheral) {
  function DigitalInput(config) {
    _classCallCheck(this, DigitalInput);

    config = parseConfig(config);
    _get(Object.getPrototypeOf(DigitalInput.prototype), "constructor", this).call(this, config.pin);
    addon.init(this.pins[0], config.pullResistor, INPUT);
    this.value = addon.read(this.pins[0]);
  }

  _inherits(DigitalInput, Peripheral);

  _prototypeProperties(DigitalInput, null, {
    read: {
      value: function read() {
        if (!this.alive) {
          throw new Error("Attempted to read from a destroyed peripheral");
        }
        this.value = addon.read(this.pins[0]);
        return this.value;
      },
      writable: true,
      configurable: true
    }
  });

  return DigitalInput;
})(Peripheral);

Object.defineProperty(exports, "__esModule", {
  value: true
});
